rfcs/proposed/host_backend_histogram/README.md

```diff
@@ -0,0 +1,155 @@
+number of bins with both OpenMP and TBB. There was some subset of cases for which the atomics implementation
+outperformed the proposed implementation (below). However, this was generally limited to some specific cases where
+the number of bins was very large (~1 Million), and even for this subset significant benefit was only found for cases
+with a small number for input elements relative to number of bins. This makes sense because the atomic implementation
+is able to avoid the overhead of allocating and initializing temporary histogram copies, which is largest when
+the number of bins is large compared to the number of input elements. With many bins, contention on atomics is also
+limited as compared to the embarassingly parallel proposal which does experience this contention.
+
+When we examine the real world utility of these cases, we find that they are uncommon and unlikely to be the important
+use cases. Histograms generally are used to categorize large images or arrays into a smaller number of bins to
+characterize the result. Cases for which there are similar or more bins than input elements are not very practical in
+practice. The maintenance and complexity cost associated with supporting and maintaining a second implementation to
+serve this subset of cases does not seem to be justified. Therefore, this implementation has been discarded at this
+time.
+
+### Other Unexplored Approaches
+* One could consider some sort of locking approach which locks mutexes for subsections of the output histogram prior to
+  modifying them. It's possible such an approach could provide a similar approach to atomics, but with different
+  overhead trade-offs. It seems quite likely that this would result in more overhead, but it could be worth exploring.
+
+* Another possible approach could be to do something like the proposed implementation one, but with some sparse
+  representation of output data. However, I think the general assumptions we can make about the normal case make this
+  less likely to be beneficial. It is quite likely that `n` is much larger than the output histograms, and that a large
+  percentage of the output histogram may be occupied, even when considering dividing the input amongst multiple
+  threads. This could be explored if we find temporary storage is too large for some cases and the atomic approach
+  does not provide a good fallback.
+
+## Proposal
+After exploring the above implementation for `histogram`, the following proposal better represents the use
+cases which are important, and provides reasonable performance for most cases.
+
+### Embarrassingly Parallel Via Temporary Histograms
+This method uses temporary storage and a pair of embarrassingly parallel `parallel_for` loops to accomplish the
+`histogram`.
+
+For this algorithm, each parallel backend will add a  `__thread_enumerable_storage<_StoredType>` struct which provides
```

Comment:
1. Why a TLS is used, not the global memory  with "thread id" as a key? F.e. `bins[thread_id]`?
2. Does TBB guarantee that the same threads finalize the work? "The same threads" means the threads which have started the work?

Reply:
1) There are a number of reasons:
a) My understanding is that in TBB while it may be technically possible to get the the local thread id within an arena, it is an undocumented API and generally discouraged and against the TBB mindset.  Using TLS seems to be the preferred method specifically with TBB.
b) While what you suggest perhaps fits better within OpenMP, we want to create a single implementation and not require a `__parallel_histogram` within every current and future backend, but rather depend upon existing functionality within the backend as much as we can (in this case `__parallel_for`).
c) With smaller values of `n`, `num_bins` and larger number of threads, not all threads *should* be used because of the overhead associated with allocation and initialization of each temporary bin copy.  We can let the partitioner decide how many blocks to employ, but we want to avoid unnecessary allocation and initialization overheads wherever possible.

I will mention a downside for completeness, but it is outweighed here in my opinion:
It requires implementation of a thread local storage class for each backend.  This is only non-trivial for OpenMP. It has been written generically though to serve future patterns though so it is nice to have.

Reply:
2) I'm not exactly sure what you mean here by "finalize the work".  If you mean the second parallel for, then no, we are explicitly parallelizing over a different dimension (`num_bins`), and accumulating across the temporary histograms which were used from different threads.  TBB does guarantee that each thread will always use its own TLS for each grain of work though, when retrieved through `local()`.